### PR TITLE
treewide: implement lazy loading via lz.n for selected plugins

### PR DESCRIPTION
**Pull request by horriblename**
things maybe to do:
- integrate into extraPlugins instead of using a separate option

bug:
- ~~there's this stupid bug where if you give only one plugin it doesn't work, I'll fix it upstream later~~ fixed

### CODE REVIEW

File: modules/plugins/debugger/nvim-dap/config.nix [hunk 1]
```diff
@@ -49,24 +49,33 @@ in {
...
     (mkIf (cfg.enable && cfg.ui.enable) {
-      vim.startPlugins = ["nvim-dap-ui" "nvim-nio"];
+      vim.startPlugins = ["nvim-nio"];
 
-      vim.pluginRC.nvim-dap-ui = entryAfter ["nvim-dap"] (''
-          local dapui = require("dapui")
-          dapui.setup()
-        ''
-        + optionalString cfg.ui.autoStart ''
+      vim.lazy.plugins = [
```
Comment (NotAShelf) [inline modules/plugins/debugger/nvim-dap/config.nix:54]: Lets not do this again please.

```nix
vim = { ... };
```

while you are changing this file already.

**Review by NotAShelf — CHANGES_REQUESTED** [review]
just being nitty at this point

haven't gone through the nix functions extensively, but from a glance they look alright - better if we test this live for a while in 0.7 after the merge

File: docs/manual/hacking/additional-plugins.md [hunk 2]
```diff
@@ -124,3 +124,61 @@ vim.your-plugin.setupOpts = {
   '';
 }
 ```
+
+## Lazy plugins {#sec-lazy-plugins}
+
+If your plugin can be lazy-loaded, you should use `vim.lazy.plugins` to add your plugin. Lazy
+plugins are managed by `lz.n`.
```
Comment (diniamo) [inline docs/manual/hacking/additional-plugins.md:131]:
```suggestion
If the plugin can be lazy-loaded, `vim.lazy.plugins` should be used to add it. Lazy
plugins are managed by `lz.n`.
```

File: docs/manual/hacking/additional-plugins.md [hunk 3]
```diff
@@ -124,3 +124,61 @@ vim.your-plugin.setupOpts = {
...
+        mode = "n";
+        action = ":YourPluginCommand";
+      }
+    ];
+  };
+;
+}
+```
+
+This results in the lua code:
```
Comment (diniamo) [inline docs/manual/hacking/additional-plugins.md:165]:
```suggestion
This results in:
```

File: modules/plugins/comments/comment-nvim/config.nix [hunk 4]
```diff
@@ -14,35 +13,30 @@ in {
...
+          mappings.toggleCurrentLine.description)
+        (mkLznExprBinding ["n"] cfg.mappings.toggleCurrentBlock ''
+            function()
+              return vim.api.nvim_get_vvar('count') == 0 and '<Plug>(comment_toggle_blockwise_current)'
+                      or '<Plug>(comment_toggle_blockwise_count)'
+            end
+          ''
+          mappings.toggleCurrentBlock.description)
+        (mkLznBinding ["x"] cfg.mappings.toggleSelectedLine "<Plug>(comment_toggle_linewise_visual)" mappings.toggleSelectedLine.description)
+        (mkLznBinding ["x"] cfg.mappings.toggleSelectedBlock "<Plug>(comment_toggle_blockwise_visual)" mappings.toggleSelectedBlock.description)
```
Comment (diniamo) [inline modules/plugins/comments/comment-nvim/config.nix:38]: I'm pretty sure
```suggestion
        (mkLznBinding ["v"] cfg.mappings.toggleSelectedLine "<Plug>(comment_toggle_linewise_visual)" mappings.toggleSelectedLine.description)
        (mkLznBinding ["v"] cfg.mappings.toggleSelectedBlock "<Plug>(comment_toggle_blockwise_visual)" mappings.toggleSelectedBlock.description)
```

File: modules/plugins/debugger/nvim-dap/config.nix [hunk 5]
```diff
@@ -52,24 +52,31 @@ in {
...
 
-        pluginRC.nvim-dap-ui = entryAfter ["nvim-dap"] (''
-            local dapui = require("dapui")
-            dapui.setup()
-          ''
-          + optionalString cfg.ui.autoStart ''
+        lazy.plugins.nvim-dap-ui = {
+          package = "nvim-dap-ui";
+          setupModule = "dapui";
+          setupOpts = {};
```
Comment (diniamo) [inline modules/plugins/debugger/nvim-dap/config.nix:60]: Do we really need to specify empty setupOpts? Also why doesn't the module have a setupOpts Nix option
```suggestion
```

File: modules/plugins/terminal/toggleterm/config.nix [hunk 6]
```diff
@@ -4,38 +4,30 @@
...
+    vim = {
+      lazy.plugins.toggleterm-nvim = {
+        package = "toggleterm-nvim";
+        cmd = ["ToggleTerm" "ToggleTermSendCurrentLine" "ToggleTermSendVisualLines" "ToggleTermSendVisualSelection" "ToggleTermSetName" "ToggleTermToggleAll"];
+        keys = [
+          (mkLznBinding ["n"] cfg.mappings.open "<Cmd>execute v:count . \"ToggleTerm\"<CR>" "Toggle terminal")
+          {
+            key = cfg.lazygit.mappings.open;
+            desc = lazygitMapDesc;
+          }
```
Comment (diniamo) [inline modules/plugins/terminal/toggleterm/config.nix:25]: Add this optionally as well.

File: modules/plugins/terminal/toggleterm/config.nix [hunk 7]
```diff
@@ -50,9 +42,9 @@ in {
             end
           })
 
-          vim.keymap.set('n', ${toJSON cfg.lazygit.mappings.open}, function() lazygit:toggle() end, {silent = true, noremap = true, desc = 'Open lazygit [toggleterm]'})
+          vim.keymap.set('n', ${toJSON cfg.lazygit.mappings.open}, function() lazygit:toggle() end, {silent = true, noremap = true, desc = '${lazygitMapDesc}'})
```
Comment (diniamo) [inline modules/plugins/terminal/toggleterm/config.nix:45]: Might as well embed the mapping string directly as well, instead of relying on toJSON (I'm not sure why we do that anywhere in the first place, when toLuaObject exists).

File: modules/plugins/utility/binds/cheatsheet/config.nix [hunk 8]
```diff
@@ -4,15 +4,18 @@
...
 }: let
   inherit (lib.modules) mkIf;
-  inherit (lib.nvim.dag) entryAnywhere;
+  inherit (lib.strings) optionalString;
 
   cfg = config.vim.binds.cheatsheet;
 in {
   config = mkIf cfg.enable {
-    vim.startPlugins = ["cheatsheet-nvim"];
+    vim.lazy.plugins.cheatsheet-nvim = {
```
Comment (diniamo) [inline modules/plugins/utility/binds/cheatsheet/config.nix:12]: It would be good to make the lazy.plugins.* attribute names consistent with the module names (check this for other modules too, I probably won't catch everything)

File: modules/plugins/utility/surround/config.nix [hunk 9]
```diff
@@ -8,25 +8,48 @@
...
+          ++ map (mkLznKey ["x"]) (with vendoredKeybinds; [visual visual_line])
+          ++ map (mkLznKey ["n"]) (with vendoredKeybinds; [
+            normal
+            normal_cur
+            normal_line
+            normal_cur_line
+            delete
+            change
+            change_line
+          ]);
```
Comment (diniamo) [inline modules/plugins/utility/surround/config.nix:49]: No. I don't even want to know what this does but figure out something better

File: modules/plugins/utility/telescope/config.nix [hunk 10]
```diff
@@ -18,46 +18,57 @@
...
-        (mkSetBinding mappings.liveGrep "<cmd> Telescope live_grep<CR>")
-        (mkSetBinding mappings.buffers "<cmd> Telescope buffers<CR>")
-        (mkSetBinding mappings.helpTags "<cmd> Telescope help_tags<CR>")
-        (mkSetBinding mappings.open "<cmd> Telescope<CR>")
-        (mkSetBinding mappings.resume "<cmd> Telescope resume<CR>")
+      lazy.plugins.telescope = {
+        package = "telescope";
+        setupModule = "telescope";
+        inherit (cfg) setupOpts;
+        # FIXME: how do I deal with extensions? set all as deps?
```
Comment (diniamo) [inline modules/plugins/utility/telescope/config.nix:27]: What do you mean by "set all as deps"? This should be fine

File: modules/wrapper/lazy/config.nix [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    (mkIf cfg.enable {
+      startPlugins = ["lz-n" "lzn-auto-require"];
+
+      optPlugins = mapAttrsToList (_: plugin: plugin.package) cfg.plugins;
+
+      luaConfigRC.lzn-load = entryBefore ["pluginConfigs"] ''
+        require('lz.n').load(${toLuaObject lznSpecs})
+      '';
+    })
+    (mkIf (!cfg.enable) {
```
Comment (diniamo) [inline modules/wrapper/lazy/config.nix:89]:
```suggestion
    })
    
    (mkIf (!cfg.enable) {
```
:trollface:

File: modules/wrapper/lazy/config.nix [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    then []
+    else [spec.keys];
+
+  notLazyConfig = concatStringsSep "\n" (mapAttrsToList specToNotLazyConfig cfg.plugins);
+in {
+  config.vim = mkMerge [
+    (mkIf cfg.enable {
+      startPlugins = ["lz-n" "lzn-auto-require"];
+
+      optPlugins = mapAttrsToList (_: plugin: plugin.package) cfg.plugins;
```
Comment (diniamo) [inline modules/wrapper/lazy/config.nix:83]: Could separate the plugin definitions to a variable since it's used for the other section as well

File: modules/wrapper/lazy/config.nix [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    (mkIf (!cfg.enable) {
+      startPlugins = mapAttrsToList (_: plugin: plugin.package) cfg.plugins;
+      luaConfigPre =
+        concatStringsSep "\n"
+        (filter (x: x != null) (mapAttrsToList (_: spec: spec.beforeAll) cfg.plugins));
+      luaConfigRC.unlazy = entryAfter ["pluginConfigs"] notLazyConfig;
+      keymaps = concatLists (mapAttrsToList specToKeymaps cfg.plugins);
+    })
+  ];
+}
```
Comment (diniamo) [inline modules/wrapper/lazy/config.nix:98]: Replace every occurance of notlazy or unlazy with diligent for lolz

File: modules/wrapper/lazy/config.nix [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+      luaConfigRC.lzn-load = entryBefore ["pluginConfigs"] ''
+        require('lz.n').load(${toLuaObject lznSpecs})
+      '';
+    })
+    (mkIf (!cfg.enable) {
+      startPlugins = mapAttrsToList (_: plugin: plugin.package) cfg.plugins;
+      luaConfigPre =
+        concatStringsSep "\n"
+        (filter (x: x != null) (mapAttrsToList (_: spec: spec.beforeAll) cfg.plugins));
+      luaConfigRC.unlazy = entryAfter ["pluginConfigs"] notLazyConfig;
```
Comment (diniamo) [inline modules/wrapper/lazy/config.nix:94]: You could also put this in pluginRC

File: modules/wrapper/rc/config.nix [hunk 15]
```diff
@@ -55,6 +55,8 @@ in {
         pluginConfigs = entryAfter ["optionsScript"] pluginConfigs;
         extraPluginConfigs = entryAfter ["pluginConfigs"] extraPluginConfigs;
         mappings = entryAfter ["extraPluginConfigs"] keymaps;
+        # FIXME: put this somewhere less stupid
+        footer = entryAfter ["mappings"] (optionalString config.vim.lazy.enable "require('lzn-auto-require').enable()");
```
Comment (diniamo) [inline modules/wrapper/rc/config.nix:59]: I think you can safely remove this comment, because this doesn't seem stupid. 2 things to do though:
- name it something other than footer
- update the _DAG entries_ manual page accordingly

File: docs/manual/hacking/additional-plugins.md [hunk 16]
```diff
@@ -124,3 +124,61 @@ vim.your-plugin.setupOpts = {
...
+  vim.lazy.plugins.your-plugin = {
+    # instead of vim.startPlugins, use this:
+    package = "your-plugin";
+
+    # if your plugin uses the `require('your-plugin').setup{...}` pattern
+    setupModule = "your-plugin";
+    inherit (cfg) setupOpts;
+
+    # events that trigger this plugin to be loaded
+    events = ["DirChanged"];
```
Comment (diniamo) [inline docs/manual/hacking/additional-plugins.md:148]:
```suggestion
    # events that trigger this plugin to be loaded
    event = ["DirChanged"];
```

File: docs/manual/hacking/additional-plugins.md [hunk 17]
```diff
@@ -124,3 +124,61 @@ vim.your-plugin.setupOpts = {
...
+This results in the lua code:
+```lua
+require('lz.n').load({
+  {
+    "name-of-your-plugin",
+    after = function()
+      require('your-plugin').setup({--[[ your setupOpts ]]})
+    end,
+
+    events = {"DirChanged"},
```
Comment (diniamo) [inline docs/manual/hacking/additional-plugins.md:174]:
```suggestion
    event = {"DirChanged"},
```

**Review by diniamo — DISMISSED** [review]

File: modules/wrapper/lazy/config.nix [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+  notLazyConfig = concatStringsSep "\n" (mapAttrsToList specToNotLazyConfig cfg.plugins);
+in {
+  config.vim = mkMerge [
+    (mkIf cfg.enable {
+      startPlugins = ["lz-n" "lzn-auto-require"];
+
+      optPlugins = mapAttrsToList (_: plugin: plugin.package) cfg.plugins;
+
+      luaConfigRC.lzn-load = entryBefore ["pluginConfigs"] ''
```
Comment (diniamo) [inline modules/wrapper/lazy/config.nix:85]: Unfortunately, this doesn't guarantee that this is placed after the theme is loaded

Comment (horriblename) [reply modules/wrapper/lazy/config.nix:85]: why would that matter?

File: lib/binds.nix [hunk 19]
```diff
@@ -67,6 +67,30 @@
...
+      key = binding.value;
+      desc = binding.description;
+    };
+
+    mkSetLuaLznBinding = binding: action: {
+      inherit action;
+      key = binding.value;
+      lua = true;
+      desc = binding.description;
+    };
```
[the remark this answers is not in the document]
Comment (horriblename) [reply lib/binds.nix:93]: nah not too late, in fact now is probably the best time to get rid of it, I don't like how it needs to import the option file either, but I didn't bother touching it. in hindsight, it should be easy to do something similar by using `options ` instead of importing the file

any specific reason you don't like this

File: modules/plugins/comments/comment-nvim/config.nix (hunk 4 above)
Comment (horriblename) [reply modules/plugins/comments/comment-nvim/config.nix:38]: the default keybind is `gc`, which would interfere with typing if we also bind it to select mode...

File: modules/plugins/utility/binds/cheatsheet/config.nix (hunk 8 above)
Comment (horriblename) [reply modules/plugins/utility/binds/cheatsheet/config.nix:12]: the vim.lazy.plugins.* is significant and must be the pname/name of the plugin package (follows how mnw names the plugin directories)

File: modules/wrapper/lazy/config.nix (hunk 18 above)
Comment (diniamo) [reply modules/wrapper/lazy/config.nix:85]: I'm not sure that it does. But having it in the correct place can avoid unforeseen issues. It's not only the theme that's loaded before, but basic and other stuff as well.

You could make this a part of the strict DAG order along with the footer.

File: lib/binds.nix (hunk 19 above)
[the remark this answers is not in the document]
Comment (diniamo) [reply lib/binds.nix:93]: Mainly because they are hard to trace/understand. I was specifically talking about these functions (and the ones for normal binds). Though the self imports might be more important to get rid of. What do you mean by `by using options instead`?

File: modules/plugins/utility/binds/cheatsheet/config.nix (hunk 8 above)
Comment (diniamo) [reply modules/plugins/utility/binds/cheatsheet/config.nix:12]: Hmm, that's not very obvious, but I can agree to it

File: modules/wrapper/lazy/config.nix (hunk 18 above)
Comment (horriblename) [reply modules/wrapper/lazy/config.nix:85]: moved lazy configs to between builtin plugins and extraPlugins + updated docs

File: lib/binds.nix (hunk 19 above)
[the remark this answers is not in the document]
Comment (horriblename) [reply lib/binds.nix:93]: too much pain, deferring this issue to the next unfortunate soul (I did change `self = import ...` to simply `options.foo` tho)

File: modules/wrapper/lazy/config.nix (hunk 18 above)
Comment (horriblename) [reply modules/wrapper/lazy/config.nix]: (jk. moved it to right before builtin plugins)

File: modules/wrapper/lazy/default.nix [hunk 20]
```diff
@@ -0,0 +1,6 @@
+_: {
```
Comment (NotAShelf) [inline modules/wrapper/lazy/default.nix:1]:
```suggestion

```

Comment (NotAShelf) [reply modules/wrapper/lazy/default.nix:1]: :D

**Review by NotAShelf — DISMISSED** [review]
LGTM

File: modules/plugins/utility/surround/config.nix (hunk 9 above)
Comment (horriblename) [reply modules/plugins/utility/surround/config.nix]: reworked the whole thing, please take another look

File: docs/manual/configuring/dag-entries.md [hunk 21]
```diff
@@ -15,9 +15,11 @@ entries in nvf:
 5. `theme` (this is simply placed before `pluginConfigs`, meaning that
    surrounding entries don't depend on it) - used to set up the theme, which has
    to be done before other plugins
-6. `pluginConfigs` - the result of the nested `vim.pluginRC` (internal option,
+6. `lazyConfigs` - `lz.n` and `lzn-auto-require` configs. If `vim.lazy.enable`
```
Comment (diniamo) [inline docs/manual/configuring/dag-entries.md:18]: Looks like raf forgot to add `optionsScript` as well, maybe you could include that

File: modules/plugins/completion/nvim-cmp/config.nix [hunk 22]
```diff
@@ -3,91 +3,122 @@
...
+      startPlugins = ["rtp-nvim"];
+      lazy.plugins = mkMerge [
+        (mapListToAttrs (package: {
+            name = getPluginName package;
+            value = {
+              inherit package;
+              lazy = true;
+              after = ''
+                local path = vim.fn.globpath(vim.o.packpath, 'pack/*/opt/${getPluginName package}')
+                require("rtp_nvim").source_after_plugin_dir(path)
```
Comment (diniamo) [inline modules/plugins/completion/nvim-cmp/config.nix:34]: Why is this needed? Why can't you simply add the source plugins to startPlugins or optPlugins?

File: modules/plugins/completion/nvim-cmp/config.nix [hunk 23]
```diff
@@ -3,91 +3,122 @@
...
-          cmp.mapping(function(fallback)
-            local has_words_before = function()
-              local line, col = unpack(vim.api.nvim_win_get_cursor(0))
-              return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
-            end
+          formatting.format = cfg.format;
+        };
+
+        # `cmp` and `luasnip` are defined above, in the `nvim-cmp` section
+        setupOpts.mapping = {
```
Comment (diniamo) [inline modules/plugins/completion/nvim-cmp/config.nix:80]: Move this into the setupOpts above

File: modules/plugins/lsp/config.nix [hunk 24]
```diff
@@ -116,7 +116,7 @@ in {
         end
 
         local capabilities = vim.lsp.protocol.make_client_capabilities()
-        ${optionalString usingNvimCmp "capabilities = require('cmp_nvim_lsp').default_capabilities()"}
+        -- ${optionalString usingNvimCmp "capabilities = require('cmp_nvim_lsp').default_capabilities()"}
```
Comment (diniamo) [inline modules/plugins/lsp/config.nix:119]: Why is this commented out?

**Review by diniamo — CHANGES_REQUESTED** [review]
Make sure to run `nix fmt` as well

File: modules/plugins/completion/nvim-cmp/config.nix (hunk 22 above)
Comment (horriblename) [reply modules/plugins/completion/nvim-cmp/config.nix:34]: cmp sources register themselves using an after/plugin/*.lua file

putting them in start plugin would make them also load cmp at startup

putting them in opt, lz.n won't source after/plugin files cuz packadd doesn't, so the sources aren't registered

File: modules/plugins/lsp/config.nix (hunk 24 above)
Comment (horriblename) [reply modules/plugins/lsp/config.nix:119]: me being dumb

this little maneuver is preventing me from lazy loading cmp so I commented it out to test things

will add it back in + lazy load lspconfig

File: modules/plugins/completion/nvim-cmp/config.nix (hunk 22 above)
Comment (diniamo) [reply modules/plugins/completion/nvim-cmp/config.nix:34]: yeah no just abolish cmp at that point

File: modules/plugins/lsp/config.nix (hunk 24 above)
Comment (horriblename) [reply modules/plugins/lsp/config.nix]: I've decided to skip on lazy-loading lspconfig since it would pretty much mean rewriting the entire language module. So cmp is now not lazy loaded

**Review by diniamo — APPROVED** [review]

**Review by NotAShelf — APPROVED** [review]
